src/modules/didIpfsHelper.js

```diff
@@ -138,26 +141,46 @@ class ChluIPFSDID {
     }
 
     async publish(did, waitForReplication = true) {
-        let publicDidDocument
+        let publicDidDocument, privateKeyBase58, signature
         if (did) { 
             publicDidDocument = did.publicDidDocument
+            signature = did.signature
+            privateKeyBase58 = did.privateKeyBase58
         } else {
             publicDidDocument = this.publicDidDocument
+            privateKeyBase58 = this.privateKeyBase58
         }
         this.chluIpfs.logger.debug(`Publishing DID ${publicDidDocument.id}, waitForReplication: ${waitForReplication ? 'yes' : 'no'}`)
-        const existingMultihash = await this.chluIpfs.orbitDb.getDID(publicDidDocument.id, false)
         const multihash = await this.chluIpfs.ipfsUtils.putJSON(publicDidDocument)
+        if (!signature) {
+            this.chluIpfs.logger.debug(`Publishing DID ${publicDidDocument.id}: missing signature, signing...`)
+            if (privateKeyBase58) {
+                // Create signature
+                signature = await this.signMultihash(multihash, { publicDidDocument, privateKeyBase58 })
+            } else {
+                throw new Error('Missing signature and private key, cannot sign Public DID Document')
+            }
+        } else {
+            this.chluIpfs.logger.debug(`Publishing DID ${publicDidDocument.id}: signature provided by caller`)
+        }
+        // Validate signature
+        const valid = await this.verifyMultihashWithDIDDocumentMultihash(multihash, multihash, signature)
```

Comment:
Kind of odd to see first two arguments with the same parameter. Can we refactor this so it is easy to read in future?

Reply:
yes this function is quite a mess, like the rest of the DID IPFS Helper module: some parts of it should really be a part of chlu-did, and the resolving and publishing should be a separate module so that the api-client can just replace the resolving and publishing and reuse the rest of it.

This refactor work is not blocked, but maybe it's best to do it when we figure out the "real" DID implementation with the resolver, what do you think?

Reply:
OK. Agreed.

Reply:
btw the multihashes are the same because I am checking that the multihash that points to the public DID Document is signed with the key that matches the public DID document itself.